File: Challenge122Easy/FSharp/Challenge122Easy.fs
// Matt Lorimor
// Language: F#
// Challenge #122 [Easy] Sum Them Digits - http://www.reddit.com/r/dailyprogrammer/comments/1berjh/040113_challenge_122_easy_sum_them_digits/
// Can be opened Visual Studio

let inputNumber = 1073741824

let subtractNine num =
    num - 9

//Uses the formula from Wikipeda for finding digital roots: http://en.wikipedia.org/wiki/Digital_root
let calculateDigitalRootModulo num =
    num % 9

//While loop version
let calculateDigitalRoot num =
    let mutable continueLoop = true
    let mutable result = subtractNine num
    while continueLoop do
        if result < 10 then
            continueLoop <- false
        else
            result <- subtractNine result
    result

//Recursive version
let rec calculateDigitalRootRecursive num =
    if num >= 10 then
        calculateDigitalRootRecursive (subtractNine num)
    else
        num

let moduloResult = calculateDigitalRootModulo inputNumber
let customResult = calculateDigitalRoot inputNumber
let recursiveResult = calculateDigitalRootRecursive inputNumber

printfn "Modulo-style returned %d." moduloResult
printfn "Custom-style returned %d." customResult
printfn "Recursive-style returned %d." recursiveResult

File: Challenge122Easy/CSharp/Challenge122Easy.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp
{
    class Challenge122Easy
    {
        static void Main(string[] args)
        {
            var inputNumber = 1073741824;
            var moduloResult = CalculateDigitalRootModulo(inputNumber);
            var customResult = CalculateDigitalRoot(inputNumber);
            Console.WriteLine(string.Format("Modulo-style returned {0} in {1} milliseconds.", moduloResult.Output, moduloResult.ElapsedMilliseconds));
            Console.WriteLine(string.Format("Custom-style returned {0} in {1} milliseconds.", customResult.Output, customResult.ElapsedMilliseconds));
        }

        //Uses the formula from Wikipeda for finding digital roots: http://en.wikipedia.org/wiki/Digital_root
        public static DigitalRootCalculation CalculateDigitalRootModulo(int num)
        {
            Stopwatch stopWatch = new Stopwatch();
            stopWatch.Start();
            DigitalRootCalculation droot = new DigitalRootCalculation(num);
            droot.Output = droot.Input % 9;
            stopWatch.Stop();
            droot.ElapsedMilliseconds = Convert.ToInt32(stopWatch.ElapsedMilliseconds);
            return droot;
        }

        //Strips the input number down until it's less than two digits
        public static DigitalRootCalculation CalculateDigitalRoot(int num)
        {
            Stopwatch stopWatch = new Stopwatch();
            stopWatch.Start();
            DigitalRootCalculation droot = new DigitalRootCalculation(num);
            droot.Output = droot.Input;
            while (droot.Output >= 10)
            {
                droot.Output -= 9;
            }
            stopWatch.Stop();
            droot.ElapsedMilliseconds = Convert.ToInt32(stopWatch.ElapsedMilliseconds);
            return droot;
        }

        public class DigitalRootCalculation
        {
            public int Input;
            public int Output;
            public int ElapsedMilliseconds;

            public DigitalRootCalculation(int num)
            {
                Input = num;
            }
        }
    }
}
